### ISSUE TITLE
Algorithms [technical]

### ISSUE THREAD

**Issue by alltheseas** [damus-io/damus#1302]
_what is it_
Algorithm capability will enable Damus users to opt-in to various algorithms as a complement to the chronological timeline.

This could enable "trending" posts, "trending" profiles etc

_implementation_
@jb55 was exploring script? language capability 

NoteID:

**Comment by GBKS** [comment]
That's a fun exploration. It might be best to start with thinking about what people might want to use algos for, and then derive the display from there. Always good to work with realistic content and scenarios. Did someone already explore that in depth? For example, is search also an algo?

Generally, the idea is that there are different views into all the activity. I like how Twitter has the row of tabs at the top when you set up multiple lists. Allows you to quickly switch between them. I also make heavy use of algo user options like muting/blocking words/people/hashtags. Twitter also has a bunch of other [search options](https://developer.twitter.com/en/docs/twitter-api/v1/rules-and-filtering/search-operators) that could be considered. For example, when searching for bitcoin content, it's often helpful to use `-filter:links` to remove spam.

It can get pretty complex technically, but I do wonder if there aren't a handful of use cases that cover 98% of all needs.

**Comment by alltheseas** [comment]
> Generally, the idea is that there are different views into all the activity. I like how Twitter has the row of tabs at the top when you set up multiple lists. Allows you to quickly switch between them.

Agree

> I also make heavy use of algo user options like muting/blocking words/people/hashtags

These are not algos, but user created lists. Unless your idea is to automate this, and have algos make these decisions on behalf of the user. This a twitter, fb very dark, not opt-in, pattern.

> Twitter also has a bunch of other [search options](https://developer.twitter.com/en/docs/twitter-api/v1/rules-and-filtering/search-operators) that could be considered. For example, when searching for bitcoin content, it's often helpful to use -filter:links to remove spam.

The above is an advanced search description. We have this on roadmap, and have not prioritized yet. https://github.com/damus-io/damus/issues/493 https://github.com/damus-io/damus/issues/1075

> It might be best to start with thinking about what people might want to use algos for
Currently the only option of timeline, and universe views is chronological. An algo could find notes in an order other than chronological. 

Top zapped replies is an example sorting. https://github.com/damus-io/damus/issues/1116
Most zapped notes view. https://github.com/damus-io/damus/issues/686

We don't need algo marketplace capability for the above two examples. However, if we are to provide "algo marketplace" capability there can be third party algo providers that experiment and provide novel algorithms for exploration. This is something that twtr, and fb are incentivized not to do, as they keep their users in a perpetual state of doomscrolling amid sale of user data, and ads bombardment.

**Comment by GBKS** [comment]
> These are not algos, but user created lists. Unless your idea is to automate this, and have algos make these decisions on behalf of the user. This a twitter, fb very dark, not opt-in, pattern.

Yes, these are put together by the user, but shouldn't an algo also consider them? Or does the algo create a feed and then the client applies the mute/block list filtering afterwards on the results? Whether I see my chronological timeline, most zapped content or do a search, I may still just not ever see certain terms or content from certain people.

In my mind, these things are all related. I create a "view into the stream of information", which involves an logic part (the algo) as well as user options (search term and parameters like filtering links, as well as the mute/block list) that the logic part works with (variables essentially).

**Comment by alltheseas** [comment]
> but shouldn't an algo also consider them?

It depends on the algo that is "installed" by the user. 

> In my mind, these things are all related. I create a "view into the stream of information", which involves an logic part (the algo) as well as user options (search term and parameters like filtering links, as well as the mute/block list) that the logic part works with (variables essentially).

Agree

**Comment by alltheseas** [comment]
https://github.com/damus-io/damus/pull/1246

**Comment by alltheseas** [comment]
User defined algos requirements:

1. Algos must be opt-in ("user enabled")
2. Algos must be known, visible, and configurable

**Comment by alltheseas** [comment]
https://damus.io/note1kpm4dhdtl3lhlmcepjgleycmrunlydjfvem5zdycmea73g8uhpmqr36z44

Algo implementation and UI via bluesky for web. UI is not normie friendly.

**Comment by alltheseas** [comment]
User feedback https://damus.io/note1tkn46rxjd73ef27cehgnqyhv0gy52ue8y3tjwkzntqnaqdln5xvs96w5dq

**Comment by jb55** [comment]
@GBKS wrote:
> Twitter also has a bunch of other search options that could be considered. For example, when searching for bitcoin content, it's often helpful to use -filter:links to remove spam.

I'm putting together a flexible way of building any kind of algorithm, filtering links would be very easy with a nostrscript + nostrdb:

```lua
-- note blocks are blocks of parsed contents
for block in nostr_note_blocks(note) {
  if block.type == url {
      return false
  }
}
return true
```

this could be assemblyscript or a custom wasm-targetted language that I'm calling nostrscripts. These filters run alongside nostrdb queries, so that are super efficient. I am also planning on creating reducers for more than just filtering.

Full plan here:

- https://github.com/damus-io/nostrdb/issues/33

**Comment by alltheseas** [comment]
discoverability is lacking

https://damus.io/nevent1qqsz28a99c5xyw7l3clu8qe3zrmatph9lckdd0tk3mg7jye9kzw33tcjrl529

**Comment by alltheseas** [comment]
Customer requests people you might like/people followed by those you follow algo

https://damus.io/nevent1qqsz8akt2lkexcunr828858xz2arn97lya60lwpd37kgkcc3m4wwqmqpz4mhxue69uhhyetvv9ujumt0wd68ytnsw43qzxrhwden5te0wfjkccte9ehx7umhdpjhyefwvdhk6qghwaehxw309ahx7um5wghxvmt59emkj73wvf5h5qgjwaehxw309ac82unsd3jhqct89ejhxr6249y

> For discovery of new people to follow, is there a ‘people you might like’ DVM / client feature out there?

> Looking for something that searches for the most followed npubs by people you already follow (only showing people you don’t already follow).

**Comment by alltheseas** [comment]
Customer request for finding valuable non-current posts

> Are there any Nostr clients that don't have such a present bias? I'd like to see older posts that have been popular and continue to get engagement as that's probably the most evergreen content.

> Wisdom > hot takes.

https://damus.io/nevent1qqsg392vpmf2gpameeh5el5ryg0qyhwev8u78y44whh3482qx4my3ncpp4mhxue69uhkummn9ekx7mqpzfmhxue69uhhqatjwpkx2urpvuhx2ucprpmhxue69uhhyetvv9ujuumwdae8gtnnda3kjctvqythwumn8ghj7un9d3shjtnrw4e8yetwwshxv7tf6zkgj0

**Comment by alltheseas** [comment]
> Customer requests people you might like/people followed by those you follow algo
> 
> https://damus.io/nevent1qqsz8akt2lkexcunr828858xz2arn97lya60lwpd37kgkcc3m4wwqmqpz4mhxue69uhhyetvv9ujumt0wd68ytnsw43qzxrhwden5te0wfjkccte9ehx7umhdpjhyefwvdhk6qghwaehxw309ahx7um5wghxvmt59emkj73wvf5h5qgjwaehxw309ac82unsd3jhqct89ejhxr6249y
> 
> > For discovery of new people to follow, is there a ‘people you might like’ DVM / client feature out there?
> 
> > Looking for something that searches for the most followed npubs by people you already follow (only showing people you don’t already follow).

@scoder1747

**Comment by alltheseas** [comment]
Closing as part of a backlog triage — we're narrowing focus to the most actionable issues to reduce noise for contributors.

If this is still relevant, please reopen with updated context or reproduction steps. Thanks for reporting!
